### PR TITLE
[FEATURE][needs-docs] Svg marker size: change aspect ratio

### PR DESCRIPTION
**Pull request by rldhont**
## Description
Qt gives us the ability to display SVG marker with a different aspect ratio than the one defined in the SVG file.
This PR updates the way to display SVG marker and gives the ability to user to update the aspect ratio.

This PR has been funded by Ville d'Avignon

## Checklist
- [ ] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/symbology-ng/qgsmarkersymbollayer.cpp
Comment (nyalldawson) [inline src/core/symbology-ng/qgsmarkersymbollayer.cpp]: Can you initialize these in the header instead of here (c++11 style)

**Review by nyalldawson — COMMENTED** [review]
Cool feature! Code looks good - just one concern I've noted above.

File: src/core/symbology-ng/qgsmarkersymbollayer.cpp
Comment (rldhont) [inline src/core/symbology-ng/qgsmarkersymbollayer.cpp]: I thought I had to do like that.

**Review by nirvn — CHANGES_REQUESTED** [review]
@rldhont , nice feature, thanks for working this out.

There's a problem with the .ui file, whereas the minimum with of the svg marker's configuration panel has dramatically expanded. 

PR looks good to me after the above .ui issue is fixed.

**Comment by rldhont** [comment]
@nirvn I have updated the ui, it is the right way ? If not, which line can I update ?

**Comment by rldhont** [comment]
@nyalldawson thanks for your review. Is my update right ?

File: tests/code_layout/acceptable_missing_doc.py
Comment (nyalldawson) [inline tests/code_layout/acceptable_missing_doc.py]: HEY! That's not how the doc test works... add documentation to the newly added members please :)

File: tests/src/core/testqgssvgmarker.cpp
Comment (nyalldawson) [inline tests/src/core/testqgssvgmarker.cpp]: This reference image looks fragile due to the ~1 pixel high areas. I'd suggest:
- remove the outline from the symbol, that'll help avoid antialiasing/blending of fill/outline platform differences
- boost the minimum size here so that the smallest symbol is at least a couple of pixels thick in its center

File: tests/src/core/testqgssvgmarker.cpp
Comment (nyalldawson) [inline tests/src/core/testqgssvgmarker.cpp]: What's happening here? is this supposed to be reusing the existing reference image? I'd have expected it to look different given that it's only the width which is varying?

**Comment by nirvn** [comment]
@rldhont , the width issue is still present:
screenshot from 2017-07-26 10-15-09

**Comment by rldhont** [comment]
@nirvn I apology but I don't understand the issue ?
I didn't see in the diff of .ui file where I have modified the minimum width of the svg marker's configuration panel. How do you want I fix it ?

**Comment by nirvn** [comment]
@rldhont , try a master vs. your PR build, and notice how the minimum width (before a scrollbar appears) of the svg marker panel in your PR is much wider than on master.

It's not clear to me which change in the .ui file led to that.

**Comment by rldhont** [comment]
@nirvn it is not just because this PR adds 2 new inputs and the panel it self is not scrollable ?

**Comment by nirvn** [comment]
@rldhont , problem has to do with the [x] preserve aspect ratio checkbox. The minimum width goes back to normal when I reduce the string to "P..." (test only, to dissect the issue):
screenshot from 2017-07-26 16-11-43

**Comment by rldhont** [comment]
@nirvn, thanks ! I'll update the .ui to fix it

**Comment by nyalldawson** [comment]
@nirvn seems like a perfect place for your new ratio lock widget?

File: tests/src/core/testqgssvgmarker.cpp
Comment (rldhont) [inline tests/src/core/testqgssvgmarker.cpp]: In this case the aspect ratio has not been explicitly changed with setFixedAspectRatio or with the PreserveAspectRatio data defined property, so the Width data defined property gives the same has the Size data defined property.
I think I have to add some tests.

File: tests/code_layout/acceptable_missing_doc.py
Comment (rldhont) [inline tests/code_layout/acceptable_missing_doc.py]: But no method has been documented here...

**Comment by rldhont** [comment]
@nyalldawson I have updated documentations and tests.
@nirvn I have updated the ui

**Comment by rldhont** [comment]
Do I rebase to reorganize commits ?

**Comment by daydinodakent** [comment]
I follow nightly edition of QGIS 2.99
There is a problem in showing point features as SVG file.
If you try to show point features as SVG marker and give any "rotation angle" or make svg marker settings in "MapUnit" then
QGIS stop working and close.

**Comment by rldhont** [comment]
@nyalldawson and @nirvn do you have any objection to merge ?

**Comment by nyalldawson** [comment]
Building now for a test...

File: src/core/symbology/qgsmarkersymbollayer.h
Comment (nyalldawson) [inline src/core/symbology/qgsmarkersymbollayer.h]: far -> ratio

File: src/core/symbology/qgsmarkersymbollayer.h
Comment (nyalldawson) [inline src/core/symbology/qgsmarkersymbollayer.h]: Since these are protected members, they'll need doc strings (or make them private)

File: src/gui/symbology/qgssymbollayerwidget.h
Comment (nyalldawson) [inline src/gui/symbology/qgssymbollayerwidget.h]: Can we make these private slots? I don't like how all these widgets expose a lot of their internal workings publicly (and hence become part of stable API). If not, they need docstrings added

**Review by nyalldawson — COMMENTED** [review]
@rldhont I just tested and it's working very (very!) nicely. My one concern is with the UI here. What I'm wondering is whether the data defined control for preserving the aspect ratio is really necessary here? If it isn't, we should use @nirvn's new QgsRatioLockButton instead of a checkbox for the ratio lock widget. This will keep consistency between different parts of the UI, and is a lot nicer than a plain checkbox.

I also think the units combo box should be vertically centered between the width and height spin boxes. At the moment it's bottom aligned, so it doesn't look associated with the width spin box.

**Comment by rldhont** [comment]
@m-kuhn, @nirvn I have updated the UI and code as you requested it.
Is it ok for you ?

**Comment by DelazJ** [comment]
Also, how about left aligning the "Preserve..." checkbox with "**H**eight" and "**W**idth" labels (capitalize first letter)? afaics, they are of the same level: subitems of the Size parameter.

**Comment by nyalldawson** [comment]
My view would still be to use the ratio luck button - it's much much less visual clutter then a checkbox with a lengthy label

**Comment by nyalldawson** [comment]
Luck=lock

**Comment by rldhont** [comment]
@m-kuhn @nirvn does the ratio lock button have documentation ? Is it already available in QGIS source tree ?

**Comment by nyalldawson** [comment]
QgsRatioLockButton. It's in master and should be suggested to use (API docs are there)

**Comment by rldhont** [comment]
Hi devs,
@nyalldawson QgsRatioLockButton added
@DelazJ labels updated
@nirvn Preserve Aspect Ratio checkbox removed
Telle me if it is good for you ?

**Comment by nirvn** [comment]
@rldhont , looking *much* better with the radio lock button:
screenshot from 2017-08-28 10-33-40

There's an issue with the ratio lock button's pressed state, whereas the state is still pressed even though the ratio lock is disabled. Check the save as image dialog to see how it should look.

Otherwise, LGTM and can't wait for it to be merged.

**Comment by rldhont** [comment]
@dmarteau can you take a look ?

**Comment by rldhont** [comment]
@nirvn I don't understand how to fix this ui point

**Comment by rldhont** [comment]
@nirvn I have updated the way Size, Height and lock button is used. But I have an issue with the setDown button, if the button is firstly displayed as locked. Can you fix this after ?

**Comment by rldhont** [comment]
@nirvn the setDown is fixed! Can I merge ?

**Comment by rldhont** [comment]
I'll merge this PR tomorrow if I've no more comments in the meantime.

**Review by nirvn — APPROVED** [review]

**Comment by nirvn** [comment]
@rldhont , sorry for the delay -- the PR is fine by me. Thanks, great feature addition.

**Pull request merged**
